### ISSUE TITLE
Still looking for static/safe async.waterfall

### ISSUE THREAD

**Issue by ORESoftware** [caolan/async#1377]
`async.waterfall` is too dynamic and frequently leads to bugs. Obviously I don't recommend changing async.waterfall - what I recommend is to add a more "static" async.waterfall function/method, let's call it `async.waterfallSafe`.

e.g. here is how dynamic async.waterfall  can be:

```js

async.waterfall([
  function(cb){
    process.next(function(){ 
          cb(null,'a','b','c');
    });
  }),

  function(a,b,c, cb){
    process.next(function(){ 
          cb(null, 'z');
    });
  }),
  function(z, cb){
    process.next(cb);
  })
], function(err, results){

 // all done

});
```

this can easily lead to bugs, because the function signatures can change too frequently - if you have branching logic in each function, than you have to be very careful to be consistent, often times passing dummy values to keep the argument count the same.

I am looking for a consistent async.waterfall API, let's call it 

`async.waterfallSafe` or `async.waterfallStatic`

it would become:

```js
async.waterfallSafe([
  function(cb){
    process.next(function(){ 
          cb(null);   // *only pass one argument, but...
    });
  }),

  function(data, cb){
    // * even though we only passed on argument, the callback is always the second arg!!!
    process.next(function(){ 
          cb(null, 'x', 'y', 'z');
    });
  }),
  function(data, cb){  // cb is always the second arg
    // data === 'z' => true
    process.next(cb);
  })
], function(err, results){

 // all done

});

```

in other words, the first function has a signature of just function(cb), but all the other functions always have a signature of function(data,cb).

This will prevent a lot of bugs and make it easier to maintain JS code that uses async.waterfall. I have seen so many bugs and headaches with this, many times with coworkers.

would you accept a PR for this functionality?

**Comment by aearly** [comment]
If `waterfall` gets too complicated, most people start using `auto`.  I don't think we need a third control flow function between `auto` and `waterfall`.

**Comment by ORESoftware** [comment]
ok I have never used auto, let me check it out

**Comment by ORESoftware** [comment]
So I checked out async.auto

http://caolan.github.io/async/docs.html#auto

auto and autoInject are both super cool, never looked into them before. As you indicated, the nice thing about these utilities is that they have a consistent signatures.

basically it's either

`function(results,cb){}`

or 

`function(cb){}`

and it's very obvious which one it should be.

thanks! should be better than async.waterfall

**Comment by aearly** [comment]
Excellent. 👍
